src/storage-client/src/controller/command_wals.rs

```diff
@@ -0,0 +1,154 @@
+    StorageResponse<T>: RustType<ProtoStorageResponse>,
+
+    Self: super::StorageController<Timestamp = T>,
+{
+    /// `true` if shard is in register for shards marked for finalization.
+    pub(super) async fn is_shard_registered_for_finalization(&mut self, shard: ShardId) -> bool {
+        SHARD_FINALIZATION
+            .peek_key_one(&mut self.state.stash, shard)
+            .await
+            .expect("must be able to connect to stash")
+            .is_some()
+    }
+
+    /// Register shards for finalization. This must be called if you intend to
+    /// finalize shards, before you perform any work to e.g. replace one shard
+    /// with another.
+    ///
+    /// The reasoning behind this is that we need to identify the intent to
+    /// finalize a shard so we can perform the finalization on reboot if we
+    /// crash and do not find the shard in use in any collection.
+    #[allow(dead_code)]
+    pub(super) async fn register_shards_for_finalization<I>(&mut self, entries: I)
+    where
+        I: IntoIterator<Item = ShardId>,
+    {
+        SHARD_FINALIZATION
+            .insert_without_overwrite(
+                &mut self.state.stash,
+                entries.into_iter().map(|key| (key, ())),
+            )
+            .await
+            .expect("must be able to write to stash")
+    }
+
+    /// Removes the shard from the finalization register.
+    ///
+    /// This is appropriate to do if you can guarantee that the shard has been
+    /// finalized or find the shard is still in use by some collection.
+    pub(super) async fn clear_from_shard_finalization_register(
+        &mut self,
+        shards: &BTreeSet<ShardId>,
+    ) {
+        SHARD_FINALIZATION
+            .delete(&mut self.state.stash, |k, _v| shards.contains(k))
+            .await
+            .expect("must be able to write to stash")
+    }
+
+    /// Reconcile the state of `SHARD_FINALIZATION_WAL` with
+    /// `super::METADATA_COLLECTION` on boot.
+    pub(super) async fn reconcile_shards(&mut self) {
+        // Get all shards in the WAL.
+        let registered_shards: BTreeSet<_> = SHARD_FINALIZATION
+            .peek_one(&mut self.state.stash)
+            .await
+            .expect("must be able to read from stash")
+            .into_iter()
+            .map(|(shard_id, _)| shard_id)
+            .collect();
+
+        if registered_shards.is_empty() {
+            return;
+        }
+
+        // Get all shards we're aware of from stash.
+        let all_shard_data: BTreeMap<_, _> = super::METADATA_COLLECTION
+            .peek_one(&mut self.state.stash)
+            .await
+            .expect("must be able to read from stash")
+            .into_iter()
+            .map(
+                |(
+                    id,
+                    super::DurableCollectionMetadata {
+                        remap_shard,
+                        data_shard,
+                    },
+                )| { [(id, [remap_shard, data_shard])] },
```

Comment:
Can you add a todo here, to also produce the schema for each shard?

Reply:
Done